internal/impl/parquet/schema_coersion.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parquet
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gofrs/uuid/v5"
+	"github.com/parquet-go/parquet-go"
+)
+
+type schemaVisitor interface {
+	visitLeaf(value any, schemaNode parquet.Node) (any, error)
+}
+
+func visitWithSchema(visitor schemaVisitor, value any, schemaNode parquet.Node) (any, error) {
+	if schemaNode.Leaf() {
+		if schemaNode.Optional() && value == nil {
+			return nil, nil
+		}
+		return visitor.visitLeaf(value, schemaNode)
+	}
+
+	switch group := value.(type) {
+	case map[string]any:
+		for _, childSchemaNode := range schemaNode.Fields() {
+			name := childSchemaNode.Name()
+			if childValue, ok := group[name]; ok {
+				var err error
+				group[name], err = visitWithSchema(visitor, childValue, childSchemaNode)
+				if err != nil {
+					return nil, fmt.Errorf("visiting [%s]: %w", name, err)
+				}
+			}
+		}
+		return group, nil
+
+	case []any:
+		for i := range group {
+			var err error
+			group[i], err = visitWithSchema(visitor, group[i], schemaNode)
+			if err != nil {
+				return nil, fmt.Errorf("visiting [%d]: %w", i, err)
+			}
+		}
+		return group, nil
+
+	case nil:
+		return nil, nil
+
+	default:
+		panic(fmt.Sprintf("unexpected group value type: %T", value))
+	}
+}
+
+type encodingCoersionVisitor struct{}
+
+func (encodingCoersionVisitor) visitLeaf(value any, schemaNode parquet.Node) (any, error) {
+	logicalType := schemaNode.Type().LogicalType()
+	if logicalType == nil {
+		return value, nil
+	}
+	if logicalType.Timestamp != nil {
+		switch v := value.(type) {
+		case string:
+			ts, err := time.Parse(time.RFC3339, v)
+			if err != nil {
+				return nil, fmt.Errorf("parsing string RFC3339 timestamp: %w", err)
+			}
+			unit := logicalType.Timestamp.Unit
+			if unit.Millis != nil {
+				return ts.UnixMilli(), nil
+			} else if unit.Micros != nil {
+				return ts.UnixMicro(), nil
+			} else if unit.Nanos != nil {
+				return ts.UnixNano(), nil
+			} else {
+				panic("unreachable")
+			}
```

Comment:
Nit: This might look better as a `switch` statement and, also, it would be best to return an error instead of having a `panic` even if it's never going to happen so we stick with the same error handling pattern.

Reply:
Turned into a switch statement